### PR TITLE
[REEF-247]: Port ranges should to be configurable for listening

### PR DESCRIPTION
**Pull request by beysims**
Defined a port range provider and a default implementation.

JIRA: [REEF-247](https://issues.apache.org/jira/browse/REEF-247)

This closes #

### CODE REVIEW

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java [hunk 1]
```diff
@@ -74,21 +75,29 @@
...
+
+  /**
    * The ConfigurationModule for the local resourcemanager.
    */
   public static final ConfigurationModule CONF = new LocalRuntimeConfiguration()
-      .merge(CommonRuntimeConfiguration.CONF)
-          // Bind the local runtime
-      .bindImplementation(JobSubmissionHandler.class, LocalJobSubmissionHandler.class)
-      .bindConstructor(ExecutorService.class, ExecutorServiceConstructor.class)
+          .merge(CommonRuntimeConfiguration.CONF)
```
Comment (markusweimer) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java]: This looks like a pure formatting change.

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java [hunk 2]
```diff
@@ -74,21 +75,29 @@
...
+                  // Bind the local runtime
+          .bindImplementation(JobSubmissionHandler.class, LocalJobSubmissionHandler.class)
+          .bindConstructor(ExecutorService.class, ExecutorServiceConstructor.class)
       .bindImplementation(RuntimeClasspathProvider.class, LocalClasspathProvider.class)
-          // Bind parameters of the local runtime
-      .bindNamedParameter(MaxNumberOfEvaluators.class, MAX_NUMBER_OF_EVALUATORS)
+              // Bind parameters of the local runtime
+          .bindNamedParameter(MaxNumberOfEvaluators.class, MAX_NUMBER_OF_EVALUATORS)
```
Comment (markusweimer) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java]: This looks like a pure formatting change.

File: lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java [hunk 3]
```diff
@@ -74,21 +75,29 @@
...
       .bindNamedParameter(RootFolder.class, RUNTIME_ROOT_FOLDER)
-      .bindNamedParameter(JVMHeapSlack.class, JVM_HEAP_SLACK)
-      .bindSetEntry(DriverConfigurationProviders.class, DRIVER_CONFIGURATION_PROVIDERS)
+          .bindNamedParameter(JVMHeapSlack.class, JVM_HEAP_SLACK)
```
Comment (markusweimer) [inline lang/java/reef-runtime-local/src/main/java/org/apache/reef/runtime/local/client/LocalRuntimeConfiguration.java]: This looks like a pure formatting change.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java [hunk 4]
```diff
@@ -82,14 +84,11 @@
         orderingGuarantee,
         numberOfTries,
         retryTimeout,
-        LocalAddressProviderFactory.getInstance());
+        LocalAddressProviderFactory.getInstance(),
+            RangeTcpPortProvider.Default);
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java]: Indent seems off.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java [hunk 5]
```diff
@@ -82,14 +84,11 @@
...
 
   }
 
-  /**
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java]: Why is this un-deprecated?

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RandomRangeIterator.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+import net.jcip.annotations.ThreadSafe;
+
+import java.util.Iterator;
+import java.util.Random;
+
+/**
+ *
+ */
+@ThreadSafe
+class RandomRangeIterator implements Iterator<Integer> {
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RandomRangeIterator.java]: make it `final`

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RandomRangeIterator.java [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+   * @return the next element in the iteration
+   */
+  @Override
+  public synchronized Integer next() {
+    return random.nextInt(tcpPortRangeCount) + tcpPortRangeBegin;
+  }
+
+  /**
+   * always throws
+   * @throws UnsupportedOperationException if the {@code remove}
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RandomRangeIterator.java]: Update the documentation of `@throws`

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RandomRangeIterator.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  private int currentRetryCount;
+  private final Random random = new Random(System.currentTimeMillis());
+
+  RandomRangeIterator(final int tcpPortRangeBegin, final int tcpPortRangeCount, int tryCount) {
+    this.tcpPortRangeBegin = tcpPortRangeBegin;
+    this.tcpPortRangeCount = tcpPortRangeCount;
+    this.tryCount = tryCount;
+  }
+
+  /**
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RandomRangeIterator.java]: If there is no change in documentation from the implemented interface, please remove the JavaDoc. When building documentation, JavaDoc is smart enough to insert the parent documentation.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RangeTcpPortProvider.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Iterator;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class RangeTcpPortProvider implements TcpPortProvider,ConfigurationProvider {
+  private final int portRangeBegin;
+  private final int portRangeCount;
+  private final int portRangeTryCount;
+  private static final Logger LOG = Logger.getLogger(RangeTcpPortProvider.class.getName());
+@Inject
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RangeTcpPortProvider.java]: Odd indent. Also, add a newline.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RangeTcpPortProvider.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.reef.wake.remote.ports.parameters.TcpPortRangeBegin;
+import org.apache.reef.wake.remote.ports.parameters.TcpPortRangeCount;
+import org.apache.reef.wake.remote.ports.parameters.TcpPortRangeTryCount;
+
+import javax.inject.Inject;
+import java.util.Iterator;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class RangeTcpPortProvider implements TcpPortProvider,ConfigurationProvider {
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RangeTcpPortProvider.java]: Make the class final and also add a JavaDoc for it.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RangeTcpPortProvider.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+   * Returns an iterator over a set of tcp ports
+   *
+   * @return an Iterator.
+   */
+  @Override
+  public Iterator<Integer> iterator() {
+    return new RandomRangeIterator (portRangeBegin, portRangeCount, portRangeTryCount);
+  }
+
+  // You should be really calling TANG
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/RangeTcpPortProvider.java]: Deprecate this and make the comment a JavaDoc to inform users what to do instead.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/TcpPortProvider.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+ */
+package org.apache.reef.wake.remote.ports;
+
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.annotations.DefaultImplementation;
+
+import java.util.Iterator;
+
+/**
+ * Gives out random tcp port numbers in a range.
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/TcpPortProvider.java]: That is not what the interface does, but the default implementation. Please update the doc.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/TcpPortProvider.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+  /**
+   * Returns an iterator over a set of tcp ports
+   *
+   * @return an Iterator.
+   */
+  @Override
+  Iterator<Integer> iterator();
+
+  /**
+   * How was this TcpPortProvider configured ?
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/TcpPortProvider.java]: Add a better JavaDoc :-)

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeBegin.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+package org.apache.reef.wake.remote.ports.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * First tcp port number to try
+ */
+@NamedParameter(doc = "First tcp port number to try", default_value = "10000")
+public class TcpPortRangeBegin implements Name<Integer> {
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeBegin.java]: Make the class `final`.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeCount.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+package org.apache.reef.wake.remote.ports.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * Number of tcp ports in the range
+ */
+@NamedParameter(doc = "Number of tcp ports in the range", default_value = "10000")
+public class TcpPortRangeCount implements Name<Integer> {
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeCount.java]: Make the class `final`.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeTryCount.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+package org.apache.reef.wake.remote.ports.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * Max number tries for port numbers
+ */
+@NamedParameter(doc = "Max number tries for port numbers", default_value = "1000")
+public class TcpPortRangeTryCount implements Name<Integer> {
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeTryCount.java]: Make the class `final`.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeTryCount.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.apache.reef.wake.remote.ports.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * Max number tries for port numbers
+ */
+@NamedParameter(doc = "Max number tries for port numbers", default_value = "1000")
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeTryCount.java]: For the default value, have it refer to the one set inside the class. That way we can never get out of sync.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeCount.java [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.apache.reef.wake.remote.ports.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * Number of tcp ports in the range
+ */
+@NamedParameter(doc = "Number of tcp ports in the range", default_value = "10000")
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeCount.java]: For the default value, have it refer to the one set inside the class. That way we can never get out of sync.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeBegin.java [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package org.apache.reef.wake.remote.ports.parameters;
+
+import org.apache.reef.tang.annotations.Name;
+import org.apache.reef.tang.annotations.NamedParameter;
+
+/**
+ * First tcp port number to try
+ */
+@NamedParameter(doc = "First tcp port number to try", default_value = "10000")
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/ports/parameters/TcpPortRangeBegin.java]: For the default value, have it refer to the one set inside the class. That way we can never get out of sync.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java [hunk 20]
```diff
@@ -139,29 +140,31 @@ public NettyMessagingTransport(final String hostAddress, int port,
...
+          } else {
+            throw ex;
           }
         }
       }
-    } catch (final Exception ex) {
-      final RuntimeException transportException =
-          new TransportRuntimeException("Cannot bind to port " + port);
-      LOG.log(Level.SEVERE, "Cannot bind to port " + port, ex);
+    }
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: I don't see what has changed here. Is it just formatting?

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java [hunk 21]
```diff
@@ -176,6 +179,26 @@ public NettyMessagingTransport(final String hostAddress, int port,
...
+     * Constructs a messaging transport
+     *
+     * @param hostAddress   the server host address
+     * @param port          the server listening port; when it is 0, randomly assign a port number
+     * @param clientStage   the client-side stage that handles transport events
+     * @param serverStage   the server-side stage that handles transport events
+     * @param numberOfTries the number of tries of connection
+     * @param retryTimeout  the timeout of reconnection
+     */
+    @Deprecated
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: The indents seem off. Also, add a `@deprecated` inside of the JavaDoc to inform users what to use instead.

**Comment by markusweimer** [comment]
I did a pass. Mostly cosmetic stuff.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java [hunk 22]
```diff
@@ -100,7 +99,11 @@
       final @Parameter(RemoteConfiguration.OrderingGuarantee.class) boolean orderingGuarantee,
       final @Parameter(RemoteConfiguration.NumberOfTries.class) int numberOfTries,
       final @Parameter(RemoteConfiguration.RetryTimeout.class) int retryTimeout,
-      final LocalAddressProvider localAddressProvider) {
+      final LocalAddressProvider localAddressProvider,
+      final TcpPortProvider tcpPortProvider) {
+
+    String fullStackTrace = org.apache.commons.lang.exception.ExceptionUtils.getFullStackTrace(new Exception());
+    LOG.log(Level.SEVERE, "zzzzz 3 " + tcpPortProvider.toString() + fullStackTrace);
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/impl/DefaultRemoteManagerImplementation.java]: Please remove this.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java (hunk 20 above)
Comment (beysims) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: yes

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java [hunk 23]
```diff
@@ -176,6 +179,27 @@ public NettyMessagingTransport(final String hostAddress, int port,
...
+     * @param hostAddress   the server host address
+     * @param port          the server listening port; when it is 0, randomly assign a port number
+     * @param clientStage   the client-side stage that handles transport events
+     * @param serverStage   the server-side stage that handles transport events
+     * @param numberOfTries the number of tries of connection
+     * @param retryTimeout  the timeout of reconnection
+     * @deprecated use the constructor that takes a TcpProvider instead
+     */
+    @Deprecated
+  public NettyMessagingTransport(final String hostAddress, int port,
```
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: Indent still looks odd.

File: lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java (hunk 20 above)
Comment (markusweimer) [inline lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/remote/transport/netty/NettyMessagingTransport.java]: Then remove the change to minimize the diff.

**Comment by markusweimer** [comment]
Still some more beauty issues.

**Pull request closed without merging**